### PR TITLE
Logging, forever support, and graceful shutdown

### PR DESCRIPTION
**Pull request by mscappini**

### CODE REVIEW

**Comment by aisthesis** [comment]
This isn't working properly. On my Ubuntu VirtualBox, it no longer even starts the server properly. There _is_ a process up, though, but it is actually owned by `root`. And it isn't writing any logs I can find.

Here's some console output for clarification:

```
marshall@ubuntu-optbot1:~/workspace/ui$ sudo npm start

> @optbot/ui@0.0.2 start /mnt/disk1/home/marshall/workspace/ui
> forever start --uid "ui" lib/app.js

warn:    --minUptime not set. Defaulting to: 1000ms
warn:    --spinSleepTime not set. Your script will exit if it does not stay up for at least 1000ms
info:    Forever processing file: lib/app.js
marshall@ubuntu-optbot1:~/workspace/ui$ ps aux | grep ui
marshall  1431  0.0  0.2  11748  2212 pts/0    S+   20:40   0:00 grep --color=auto ui
marshall@ubuntu-optbot1:~/workspace/ui$ ps aux | grep forever
root      1428  0.8  3.1 602120 32192 ?        Ss   20:40   0:00 /usr/bin/nodejs /usr/lib/node_modules/forever/bin/monitor lib/app.js
marshall  1433  0.0  0.2  11748  2192 pts/0    S+   20:40   0:00 grep --color=auto forever
marshall@ubuntu-optbot1:~/workspace/ui$
```

At this point, although the process is running, I can't hit the server from my browser, using `http://192.168.56.101:8080/`.

Moreover, the logs aren't getting written: 

```
marshall@ubuntu-optbot1:~/workspace/ui$ ll /var/log | grep quichean
marshall@ubuntu-optbot1:~/workspace/ui$ ll /mnt/disk1/var/log/
mongodb/ optbot/  
marshall@ubuntu-optbot1:~/workspace/ui$
```

I had some trouble making progress on these requirements with `quotepuller` last night, and was looking at the npm package [start-stop-daemon](https://www.npmjs.com/package/start-stop-daemon), which wraps and presupposes `forever` and might help to solve some of our problems.

File: README.md [hunk 1]
```diff
@@ -29,6 +29,10 @@ To change the bound host
 
 	$ sudo npm config set @optbot/ui:host 127.0.0.1 --global
 
+###Dependencies
```
Comment (aisthesis) [inline README.md]: The README should also include instructions for stopping the service.

File: package.json [hunk 2]
```diff
@@ -23,12 +24,14 @@
     "views/"
   ],
   "config": {
+    "logfile": "/var/log/quichean/ui/ui.log",
```
Comment (aisthesis) [inline package.json]: Let's stick with the `optbot` path for naming our logs, as that has already been started. Morevoer, for consistency with other services, we should have here `/var/log/optbot/ui/service.log`.

**Comment by mscappini** [comment]
If you could do me a favor to help troubleshoot, that'd be useful.

Edit the package.json and change the `npm start` to this:

```
forever -l ./ui-l.log -e ./ui-e.log -o ./ui-o.log start --uid \"ui\" lib/app.js
```

That will cause `forever` to write logs to the respective stream/file. If any exist and have content, is there anything useful?

**Comment by aisthesis** [comment]
troubleshooting via a machine that you're not on with commands that someone else has to enter and then providing results via dropbox or comments here isn't a viable workflow. it can't possibly end with just that and ends up taking 2 people a high multiple of the time it would take 1 to do it directly. can't you run it yourself on a virtual machine?

**Comment by mscappini** [comment]
Really? It _is_ a viable workflow. That's how people get help on projects everyday. It _does_ work on my virtual machine. That's why I suspect it's something with your configuration. That's why I'm asking you for logs. This is how troubleshooting works.

**Comment by aisthesis** [comment]
it works for you under Ubuntu?

**Comment by aisthesis** [comment]
The workflow that is not in general viable is: The code under is developed under Windows but untested under Ubuntu, which is the OS we're using in production. I'm happy to debug if it's just a question of "It works on my Ubuntu 14.04 configured as in the wiki, why isn't it working on yours?". But I don't think that's what we're dealing with here. If you really think it's just a few wrinkles that need to be ironed out, let's set a time limit where I can get back to working on my own services if it's not yet running on Ubuntu as a background process owned by user 'ui' after `x` number of hours.

Just to illustrate how I've done this: I've developed much of the code I've checked in on my Mac, but prior to submitting a pull request, I've tested it on an Ubuntu server that simulates as closely as possible the production server. I have the code working on Ubuntu before submitting the pull request. There may still be issues when I deploy to production, but that keeps them to a minimum.

If we had a Windows server, I wouldn't expect you to figure out how to make it work there if I had developed and tested my code only on OS X or Linux, and I don't feel like it's fair for you to expect me to put in a lot of hours to bridge the gap between Windows and Ubuntu every time it makes a difference what OS the code is running on.

**Comment by aisthesis** [comment]
This isn't actually shutting the process down. I tried this a few times, and on `CTRL+C` it blocks the console window for a moment, then when I hit enter, I get the shell back, but the process clearly hasn't actually terminated. From another shell, I got this result:

```
$ ps aux | grep app
root      1447  0.0  0.0   4452   688 pts/2    S+   22:09   0:00 sh -c node lib/app.js
root      1448  0.1  4.4 903624 45288 pts/2    Sl+  22:09   0:00 node lib/app.js
root      1466  0.0  0.0   4452   764 pts/0    S    22:12   0:00 sh -c node lib/app.js
root      1467  0.8  4.5 906528 46740 pts/0    Sl   22:12   0:00 node lib/app.js
root      1506  0.0  0.0   4452   856 pts/0    S+   22:13   0:00 sh -c node lib/app.js
root      1507  1.4  4.2 608396 43604 pts/0    S+   22:13   0:00 node lib/app.js
marshall  1594  0.0  0.2  11748  2276 pts/3    S+   22:14   0:00 grep --color=auto app
marshall@ubuntu-optbot1:~$
```

One of these may be `restapi`, which is running on the same server. But it's actually a problem that you can't tell the difference just looking at the process. I guess that will be resolved when we have them running as their own users. But maybe it would be better also to name the process after the service--instead of `app.js`, maybe `ui.js`, etc. I'll see what happens when I send `SIGTERM` with kill, but it's otherwise hard to know just what you're killing.

**Comment by aisthesis** [comment]
`SIGTERM` does shut it down correctly. Not sure why it isn't picking up `SIGINT`.

**Comment by aisthesis** [comment]
Just to be clear, here's the behavior I'm getting for `CTRL+C`, which should be sending `SIGINT`to the process:

```
$ sudo npm start

> @optbot/ui@0.0.2 start /mnt/disk1/home/marshall/workspace/ui
> node lib/app.js

Express server started at http://:::8080
^Cmarshall@ubuntu-optbot1:~/workspace/ui$ Received kill signal, shutting down gracefully.
Server shutting down...
```

But the shell then blocks until you hit `ENTER`. Weird. Maybe we should just leave out `SIGINT`, since it will be unavailable anyway when the process is running as daemon.

**Comment by aisthesis** [comment]
That fixes the problem, although I'm surprised that there's not a way to make this happen without creating an additional package. Did you try `socket.end()`, as I suggested above?

**Comment by aisthesis** [comment]
The log file in the `package.json` also still needs to be changed, as noted above. It generates an error as it stands.

**Comment by mscappini** [comment]
In order to do `socket.end()`, I'd have to track all of the sockets that connect and have the `connection` header set to `Keep-Alive`. Because I would have no way to close the sockets from the server side without them coming back for another request. I figured that would be more of a waste to track them. But maybe it wouldn't, I'm not really sure at the moment.

The logfile config setting should just be `./service.log`?

**Comment by mscappini** [comment]
I also kinda like the module approach because I will use it in our other Express apps.

**Comment by aisthesis** [comment]
On logfile, yes: `"./service.log"`

**Comment by aisthesis** [comment]
I certainly agree with modular. My concern, though, is that this has to be something that happens every time you run an express server on Linux. So it seems weird that we would need this. No one else is shutting it down gracefully? Or is it in our case possibly because the connection with `restapi` has to be force closed?

**Comment by Irialad** [comment]
This is why it should be discussed. There's something wrong here.

**Comment by mscappini** [comment]
Well, several modules do exist for a "graceful shutdown." So I would assume that is the case. I'm guessing most people just don't care if it's graceful or not. E.g. If you're going down, just shut the server down. Shutting down _gracefully_ is better for zero downtime implementations on a load balanced network, though, which is not something that Express caters to immediately.

**Comment by aisthesis** [comment]
Ship it! I think we addressed the outstanding questions in person, but then I forgot to approve the merge subsequently. We may want to polish the `shutdown-express` module later to make it more configurable, but we're good for this iteration.

**Pull request merged**